pkg/scheduler/queue/queue.go

```diff
@@ -443,8 +443,14 @@ func (q *RequestQueue) WaitForRequestForQuerier(ctx context.Context, last Tenant
 	select {
 	case q.waitingQuerierConns <- waitingConn:
 		select {
-		case reqForQuerier := <-waitingConn.recvChan:
+		case reqForQuerier, ok := <-waitingConn.recvChan:
+			if !ok {
+				// Channel was closed without having a value written to it. This can happen if
+				// the write context was canceled.
+				return nil, last, ErrStopped
```

Comment:
at this point we just think that a stop has caused this, returning ErrStopped may not be accurate.

I think it might be better to fix this is in `send`

Right not send looks like this:

```golang
func (wqc *waitingQuerierConn) send(req requestForQuerier) bool {
	defer close(wqc.recvChan)

	select {
	case wqc.recvChan <- req:
		return true
	case <-wqc.querierConnCtx.Done():
		return false
	}
}
```

but the select statement waiting on this end right below here is also holding a reference to the context and waiting for a `Done()`. When it experiences this, it returns `return nil, last, ctx.Err()` - I think it's preferable to his that case and let the  ctx.Err() get returned.

```golang
func (wqc *waitingQuerierConn) send(req requestForQuerier) bool {
	select {
	case wqc.recvChan <- req:
        close(wqc.recvChan)
		return true
	case <-wqc.querierConnCtx.Done():
        // do not close channel here;
        // waiting goroutine in WaitForRequestForQuerier will receive the context cancel and handle it
        // simply notify the caller in trySendNextRequestForQuerier that we did not successfully send a request
		return false
	}
}
```

I think the reason this has shown up so rarely is that precisely this, that the receiver here is also waiting on select case of context.Done() with the same context. Select branches are randomized so you ended up in a case rare case where both are true - the channel has been closed without anything written, and the context is canceled, but we handled the channel-close-read-nil case before we handled the context cancel

Reply:
Cool. I was going to ask you about that, as it wasn't clear what error to return. Does closing `recvChan` still have value if we change `send` so send+send are always paired?

Reply:
yeah I guess we don't really need to close it. still worth maybe including a comment about why we don't close it.

There's definitely a maybe cleaner alternate implementation here where instead of both sides checking the same context cancel, only the `send` side checks it context, then it sends the `ctx.Err()` to an error channel.
That way the receiving side can wait on either the recvChan or the errChan instead of checking either recvChan or context cancel. The reason we don't use an errChan right now is that we always want to send the `lastTenantIndex` back no matter what - that might go away in the future but either way probably not worth fiddling with for now.

Reply:
Cool. Let me know how that looks.

> There's definitely a maybe cleaner alternate implementation

The way that `WaitForRequestForQuerier` looks directly at `ctx.Done()` and gives up when that is closed definitely has clarity going for it. Any time things are plumbed through multiple channels I have to start drawing pictures. But, I'm not very intimate with this particular code so you might be right.